File: application/src/test/scala/com/yahoo/application/ApplicationBuilderTest.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application

import container.JDiscTest._
import java.nio.file.Files
import org.junit.Assert.{assertTrue, assertThat, fail}
import org.junit.Test
import com.yahoo.io.IOUtils
import org.hamcrest.CoreMatchers.containsString

/**
 * @author tonytv
 */
 class ApplicationBuilderTest {
    @Test
    def query_profile_types_can_be_added() {
        withApplicationBuilder { builder =>
            builder.queryProfileType("MyProfileType",
              <query-profile-type id="MyProfileType">
                <field name="age" type="integer" />
                <field name="profession" type="string" />
                <field name="user" type="query-profile:MyUserProfile" />
              </query-profile-type>)

          assertTrue(Files.exists(builder.getPath.resolve("search/query-profiles/types/MyProfileType.xml")))
        }
    }


  @Test
  def query_profile_can_be_added() {
    withApplicationBuilder { builder =>
      builder.queryProfile("MyProfile",
        <query-profile id="MyProfile">
          <field name="message">Hello world!</field>
        </query-profile>)

      assertTrue(Files.exists(builder.getPath.resolve("search/query-profiles/MyProfile.xml")))
    }
  }

  @Test
  def rank_expression_can_be_added() {
     withApplicationBuilder { builder =>
       builder.rankExpression("myExpression", "content")
       assertTrue(Files.exists(builder.getPath.resolve("searchdefinitions/myExpression.expression")))
    }
  }

  @Test
  def builder_cannot_be_reused() {
    val builder = new ApplicationBuilder
    builder.servicesXml(<jdisc version="1.0" />)

    using(builder.build()) { builder => }

    try {
      builder.servicesXml("")
      fail("Expected exception.")
    } catch {
      case e: RuntimeException => assertThat(e.getMessage, containsString("build method"))
    }

  }

  def withApplicationBuilder(f: ApplicationBuilder => Unit) {
    val builder = new ApplicationBuilder()
    try {
      f(builder)
    } finally {
      IOUtils.recursiveDeleteDir(builder.getPath.toFile)
    }
  }
}

File: application/src/test/java/com/yahoo/application/ApplicationBuilderTest.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application;

import com.yahoo.io.IOUtils;
import org.junit.jupiter.api.Test;

import java.nio.file.Files;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * @author Tony Vaagenes
 * @author ollivir
 */
public class ApplicationBuilderTest {
    @Test
    void query_profile_types_can_be_added() throws Exception {
        withApplicationBuilder(builder -> {
            builder.queryProfileType("MyProfileType", "<query-profile-type id=\"MyProfileType\">" + //
                    "<field name=\"age\" type=\"integer\" />" + //
                    "<field name=\"profession\" type=\"string\" />" + //
                    "<field name=\"user\" type=\"query-profile:MyUserProfile\" />" + //
                    "</query-profile-type>");

            assertTrue(Files.exists(builder.getPath().resolve("search/query-profiles/types/MyProfileType.xml")));
        });
    }

    @Test
    void query_profile_can_be_added() throws Exception {
        withApplicationBuilder(builder -> {
            builder.queryProfile("MyProfile",
                    "<query-profile id=\"MyProfile\">" +
                            "<field name=\"message\">Hello world!</field>" +
                            "</query-profile>");

            assertTrue(Files.exists(builder.getPath().resolve("search/query-profiles/MyProfile.xml")));
        });
    }

    @Test
    void rank_expression_can_be_added() throws Exception {
        withApplicationBuilder(builder -> {
            builder.rankExpression("myExpression", "content");
            assertTrue(Files.exists(builder.getPath().resolve("schemas/myExpression.expression")));
        });
    }

    // application unreferenced inside try
    @Test
    @SuppressWarnings("try")
    void builder_cannot_be_reused() throws Exception {
        Throwable exception = assertThrows(RuntimeException.class, () -> {

            ApplicationBuilder builder = new ApplicationBuilder();
            builder.servicesXml("<container version=\"1.0\" />");
            try (Application application = builder.build()) {
                // do nothing
            }

            builder.servicesXml(""); // should fail
        });
        assertTrue(exception.getMessage().contains("build method")); // should fail
    }

    private interface TestCase {
        void accept(ApplicationBuilder ab) throws Exception;
    }

    private static void withApplicationBuilder(TestCase f) throws Exception {
        ApplicationBuilder builder = new ApplicationBuilder();
        try {
            f.accept(builder);
        } finally {
            IOUtils.recursiveDeleteDir(builder.getPath().toFile());
        }
    }
}
